gelf/reader.go

```diff
@@ -64,67 +75,51 @@ func (r *Reader) Read(p []byte) (int, error) {
 func (r *Reader) ReadMessage() (*Message, error) {
 	cBuf := make([]byte, ChunkSize)
 	var (
-		err        error
-		n, length  int
-		cid, ocid  []byte
-		seq, total uint8
-		cHead      []byte
-		cReader    io.Reader
-		chunks     [][]byte
+		err       error
+		n         int
+		chunkHead []byte
+		cReader   io.Reader
 	)
+	var message []byte
 
-	for got := 0; got < 128 && (total == 0 || got < int(total)); got++ {
+	for {
 		if n, err = r.conn.Read(cBuf); err != nil {
 			return nil, fmt.Errorf("Read: %s", err)
 		}
-		cHead, cBuf = cBuf[:2], cBuf[:n]
-
-		if bytes.Equal(cHead, magicChunked) {
-			// fmt.Printf("chunked %v\n", cBuf[:14])
-			cid, seq, total = cBuf[2:2+8], cBuf[2+8], cBuf[2+8+1]
-			if ocid != nil && !bytes.Equal(cid, ocid) {
-				return nil, fmt.Errorf("out-of-band message %v (awaited %v)", cid, ocid)
-			} else if ocid == nil {
-				ocid = cid
-				chunks = make([][]byte, total)
+		chunkHead, cBuf = cBuf[:2], cBuf[:n]
+		if bytes.Equal(chunkHead, magicChunked) {
+			messageID := string(cBuf[2 : 2+8])
+			assembler := getAssembler(r, messageID)
+			if assembler.processed >= MaxChunksCount {
+				return nil, fmt.Errorf("too many chunks count for messageID: %v (max count %v)", messageID, MaxChunksCount)
 			}
-			n = len(cBuf) - chunkedHeaderLen
-			// fmt.Printf("setting chunks[%d]: %d\n", seq, n)
-			chunks[seq] = append(make([]byte, 0, n), cBuf[chunkedHeaderLen:]...)
-			length += n
-		} else { // not chunked
-			if total > 0 {
-				return nil, fmt.Errorf("out-of-band message (not chunked)")
+			if assembler.Expired() {
+				return nil, fmt.Errorf("message with ID: %v is expired", messageID)
 			}
-			break
-		}
-	}
-	// fmt.Printf("\nchunks: %v\n", chunks)
-
-	if length > 0 {
-		if cap(cBuf) < length {
-			cBuf = append(cBuf, make([]byte, 0, length-cap(cBuf))...)
-		}
-		cBuf = cBuf[:0]
-		for i := range chunks {
-			// fmt.Printf("appending %d %v\n", i, chunks[i])
-			cBuf = append(cBuf, chunks[i]...)
+			if messageCompleted := assembler.Update(cBuf); !messageCompleted {
+				continue
+			}
+			message = assembler.Bytes()
+			chunkHead = message[:2]
+			delete(r.chunkAssemblers, messageID)
```

Comment:
This feels like it needs a lock as otherwise it could create a problem with `getAssember`/`cleanupExpiredAssemble`

```suggestion
			delete(r.chunkAssemblers, messageID)
```

Reply:
oh, yep, forgot to add lock here